[src] Unpin host memory, fix race in cuda-online-pipeline

kkm000 opened the pull request:
* Unpin memory of h_all_waveform_, deallocated but left pinned upon
  instance destruction.
* Add a destructor to the class, and invoke WaitForLatticeCallbacks()
  to prevent a short race between the destorying of this instance and
  completion of callback threads, which decrement the pending callback
  count _after_ the callback has returned and been deleted.
* Add assertions to the destructor.

@galv, @hugovbraun, PTAL.

kkm000 commented on src/cudadecoder/batched-threaded-nnet3-cuda-online-pipeline.cc:
```diff
@@ -43,9 +43,10 @@ BatchedThreadedNnet3CudaOnlinePipeline::~BatchedThreadedNnet3CudaOnlinePipeline(
+  // n_lattice_callbacks_not_done_ is accessed after a callback has returned.
   WaitForLatticeCallbacks();
   KALDI_ASSERT(n_lattice_callbacks_not_done_ == 0);
-  KALDI_ASSERT(!cuda_fst_ || cuda_fst_.use_count() == 1);
```
In the latest version (#4548), I changed it to unique_ptr, so the count is no more. It was only passed to the decoder, exactly once, anyway, and as a raw pointer.

kkm000 replied:
Let me pre-squash separate commits on this branch and push it clean, so that line like this aren't confusing. I added and removed the assert, because in the other PR #4548 I replaced the shared_ptr with a unique_ptr.

hugovbraun commented:
Looks good. Good point for invoking WaitForLatticeCallbacks() in the destructor - this becomes even more necessary in a new commit I'm about to push, where more attributes from the class are used in the callback threads.

kkm000 commented:
@hugovbraun. thanks for the review. `WaitForLatticeCallbacks()` in the destructor is a dangerous thing. We need a cleaner solution—which may be just documenting that the user should not rely on it, and always call the function explicitly (I do, _after_ having run into that race :) ).  I Doxygenate the parts that I touch, hoping to build up to a full documentation of the CUDA decoders eventually (http://kaldi-asr.org/doc/ is built entirely from it, and doesn't yet mention them).

Maybe it's also a good idea to print a warning if the callback count is non-zero at entry into the destructor. What do you think?

The race wasn't quite obvious to the eye: I deleted the object after all my lattice callbacks were done, but the pool thread spent additional cycles in `delete` of the callback object, and by the time it reached the atomic decrement at line 628, there was no `this` anymore and nothing to decrement. ASAN is a godsend.

https://github.com/kaldi-asr/kaldi/blob/f34d4f4a92647d5cc92552572f6b4f1740b81a9f/src/cudadecoder/batched-threaded-nnet3-cuda-online-pipeline.cc#L622-L629

But by adding `WaitForLatticeCallbacks()` to the destructor, I introduced a potential deadlock. If the user attempts to delete the object from a lattice callback, `WaitForLatticeCallbacks()` will loop forever, because the destructor would have been called on the stack of a call started at line 624, and the decrement of `n_lattice_callbacks_not_done_`  at the line 628 will have never be reached.

I think that moving the line 628 above 623 should resolve this issue (as long as the user won't try to access the pipeline from the callback, of course—need to document that, too). Both `callback` and `dlat` are on the stack. . Do you see any flaws in my reasoning? I think the lattice callback is the only external part of the interface prone to this problem.

---

Deleting `CudaOnlinePipelineDynamicBatcher` from a lattice callback is a whole different story: the thread attempting to stop the thread pool belongs to the thread pool itself, so the thread will attempt to stop the thread pool, ending up calling `join` on itself. Fortunately, C++11 requires a deterministic behavior in such a case, namely throwing the exception `std::system_error(std::errc::resource_deadlock_would_occur)`. As long as we're not in the UB land _(illac sunt dracones)_, I'm good. :)

But perhaps allowing the user to construct the thread pool, own it and loan to the `CudaOnlinePipelineDynamicBatcher` w/o ownership transfer would make the whole rig more flexible. The whole kaboodle is a bit more monolithic than the rest of Kaldi's codebase, and at times hard to productize.
